File: app/src/main/java/com/mehmetakiftutuncu/muezzin/utilities/RemainingTime.java
package com.mehmetakiftutuncu.muezzin.utilities;

import org.joda.time.LocalTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

/**
 * Created by akif on 17/09/16.
 */
public final class RemainingTime {
    public static final DateTimeFormatter FORMATTER                 = DateTimeFormat.forPattern("HH:mm:ss");
    public static final DateTimeFormatter FORMATTER_WITHOUT_SECONDS = DateTimeFormat.forPattern("HH:mm");

    private static final int HOURS_IN_DAY      = 24;
    private static final int MINUTES_IN_HOUR   = 60;
    private static final int SECONDS_IN_MINUTE = 60;
    private static final int SECONDS_IN_HOUR   = 3600;

    public static LocalTime to(LocalTime to) {
        LocalTime now = LocalTime.now();

        int toSeconds            = (to.getHourOfDay() * SECONDS_IN_HOUR)  + (to.getMinuteOfHour() * 60)  + to.getSecondOfMinute();
        int nowSeconds           = (now.getHourOfDay() * SECONDS_IN_HOUR) + (now.getMinuteOfHour() * 60) + now.getSecondOfMinute();
        int nowToMidnightSeconds = ((HOURS_IN_DAY - now.getHourOfDay() - 1) * SECONDS_IN_HOUR) + ((MINUTES_IN_HOUR - now.getMinuteOfHour() - 1) * 60) + (SECONDS_IN_MINUTE - now.getSecondOfMinute() - 1);

        int diffSeconds = (toSeconds > nowSeconds) ? (toSeconds - nowSeconds) : (toSeconds + nowToMidnightSeconds);

        int seconds = diffSeconds % SECONDS_IN_MINUTE;
        int minutes = ((diffSeconds - seconds) / SECONDS_IN_MINUTE) % SECONDS_IN_MINUTE;
        int hours   = ((diffSeconds - (SECONDS_IN_MINUTE * minutes) - seconds) / SECONDS_IN_HOUR) % HOURS_IN_DAY;

        return new LocalTime(hours, minutes, seconds, 0);
    }
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/utilities/RemainingTime.kt
package com.mehmetakiftutuncu.muezzin.utilities

import org.joda.time.LocalTime
import org.joda.time.format.DateTimeFormat
import org.joda.time.format.DateTimeFormatter

object RemainingTime {
    val formatter: DateTimeFormatter               = DateTimeFormat.forPattern("HH:mm:ss")
    val formatterWithoutSeconds: DateTimeFormatter = DateTimeFormat.forPattern("HH:mm")

    private const val hoursInDay      = 24
    private const val minutesInHour   = 60
    private const val secondsInMinute = 60
    private const val secondsInHour   = 3600

    fun to(to: LocalTime): LocalTime {
        val now = LocalTime.now()

        val toSeconds  = to.hourOfDay * secondsInHour + to.minuteOfHour * 60 + to.secondOfMinute
        val nowSeconds = now.hourOfDay * secondsInHour + now.minuteOfHour * 60 + now.secondOfMinute

        val nowToMidnightSeconds = (hoursInDay - now.hourOfDay - 1) * secondsInHour + (minutesInHour - now.minuteOfHour - 1) * 60 + (secondsInMinute - now.secondOfMinute - 1)

        val diffSeconds = if (toSeconds > nowSeconds) toSeconds - nowSeconds else toSeconds + nowToMidnightSeconds

        val seconds = diffSeconds % secondsInMinute
        val minutes = (diffSeconds - seconds) / secondsInMinute % secondsInMinute
        val hours   = (diffSeconds - secondsInMinute * minutes - seconds) / secondsInHour % hoursInDay

        return LocalTime(hours, minutes, seconds, 0)
    }
}